### DISCUSSION TITLE
Padding in dp

### DISCUSSION THREAD

**Discussion by DasLixou** [AvaloniaUI/Avalonia#11691] [Q&A] [answered]
Hello!
I'm trying to implement some picked elements from Material 3 for Avalonia but in Avalonia things like padding have values which get represented by pixel if I am correct. Is there a way to set padding to a dp value?
Thanks :D

**Comment by jp2masa** [comment, marked as answer]
Values like padding, margin,... are not by pixel. They scale with dpi, although not using the same scale as dp units.

**Reply by DasLixou** [reply]
So Padding="10" already is 10dp? Thanks :D

**Reply by robloo** [reply]
As stated Avalonia's device-independent pixels are different than "dp". However, both are NOT physical pixels on a screen.

As I understand it, Avalonia's pixels should generally be equal to 1/96 of an inch on the physical device. That means each pixel is assumed to be on a 96 dpi display. Avalonia behind the scenes should do all the scaling between the actual screen resolution and Avalonia's pixel units to make the math work out. Someone with more knowledge in this area will have to correct me if I'm wrong though.

This doesn't work quite the same as in WPF though from what I've seen, especially with bitmaps. And then when a desktop or device scaling factor is applied the math changes again.

Historically, I think with WPF the intention was setting 10 device-independent pixels would be the same physical size (if you held up a ruler) on any screen. In practice I don't think that ever worked especially with system-wide scaling factors. Moreover, you don't actually want to do that as 10px on a tiny screen looks completely different than 10px on an 8K TV. So all of this is just an approximation and is designed to "just look right".

Rest assured, setting 10px in Avalonia will NOT be 10 physical pixels though and the system and Avalonia will scale things to make it look proper everywhere.

**Reply by DasLixou** [reply]
So when Material spec say I should make my sidebar 80dp then setting Width="80" is correct?

**Reply by robloo** [reply]
I tried to give more explanation to help you understand but @jp2masa already tells you they aren't the same thing. That said, yes, just set `Width="80"` it will be fine...

**Reply by DasLixou** [reply]
Oh wait... I reread this conversation and now understand it 😅 sorry for that. So now I have a problem because my icon and fontsize don't match to the size of the container and did some maths... When a Avalona "Pixel" as @robloo defined it is 1/96 of an inch on the physical device and a dp is 1/160 of an inch on the physical device it should just be dp\*160/96 = 133.33.. which is even bigger.. when i do it otherwise so dp\*96/160 it is 48 which is too small. So I am now confused... what is the right math to convert from dp to avalonias unit?

**Reply by DasLixou** [reply]
updates?

**Reply by robloo** [reply]
Without me digging into rendering code, someone more knowledgeable in this area will have to give the equation for conversion. I suspect it's non-trivial though.

**Reply by jp2masa** [reply]
I think it's always `scaling * 96` dpi. For desktop platforms, this corresponds to screen scale (typically 100% or 125%). I don't know about wasm and iOS, but for Android, the dpi is `DisplayMetrics.Density` ([Avalonia source](https://github.com/AvaloniaUI/Avalonia/blob/124f124617ffcd07da78e980ed04c0e6df595102/src/Android/Avalonia.Android/Platform/SkiaPlatform/InvalidationAwareSurfaceView.cs#L82), [Android docs](https://developer.android.com/reference/android/util/DisplayMetrics)), which is actually scaled to 160 dpi, not 96 dpi, but testing on real devices it looked closer to a 1:1 relation.
